Ready/sd 1062

cryogenian opened the pull request:
Smoke tests for SD-962, partially resolves SD-1062 (it doesn't test directory/zip because we have no zip lib yet)
I'll squash after https://github.com/slamdata/purescript-webdriver/pull/38 merge.

@jonsterling Please, review

jonsterling commented on test/src/Test/Selenium/File.purs:
```diff
@@ -765,6 +771,156 @@ moveDeleteFile = do
+
+  tryRepeatedlyTo do
+    handles <- getAllWindowHandles
+    switchTo $ fromJust $ handles !! 1
```
Should we replace these partial function calls with error handlers that bubble up notices in the tests?

I don't have a strong opinion, but I thought I'd mention it.

cryogenian replied:
The situation where `handles` have less then two elements is impossible because `proceedDownload` will popup new window. 
Not sure about other `fromJust` I can replace them if you like

jonsterling replied:
It's fine as-is, I think. But would you mind putting in a comment that says something like, "This is benign" or something? Just so that the assumptions are explicit for someone looking at the code later on. thanks!

cryogenian replied:
Ok

jonsterling commented on bower.json:
```diff
@@ -40,6 +40,7 @@
     "purescript-markdown": "garyb/purescript-markdown#021749b26d4a6683514d39110779938b7ff8df7e",
     "purescript-markdown-halogen": "^0.5.0",
     "purescript-minimatch": "~0.2.0",
+    "purescript-node-fs-aff": "~0.1.2",
```
Would a `^` bound be OK here?

cryogenian replied:
oops, will change, sorry

cryogenian commented:
Travis fails because lack of memory. On my laptop compilation takes exactly 3Gb RAM.

jonsterling commented:
@cryogenian Oh no...

jonsterling commented:
@cryogenian What do you recommend we do? I really don't want to get into a situation where Travis is not working for any period of time; last time, it was kind of devastating.

cryogenian commented:
We can make tests separate project. Removing `test/src/**/*.purs` from list of sources reduce memory consumption by 400Mb. Probably psc 0.7.4 eats less memory.

jonsterling commented:
@cryogenian Interesting idea... If we did make the tests a separate project though, would there still be a way to have them run on every PR?

Also, do you think that this PR has somehow pushed memory consumption over the edge, or that it would have happened to any PR at this point?

cryogenian commented:
I think we can clone test project in travis. And main project in test's travis. 

Don't know. Builds continues to become slower with every pr probably, but it's only assumption, I've not tested it precisely

jonsterling commented:
OK... @garyb do you have any thoughts here?

I'm kind of reluctant to merge a PR that will result in our travis builds being broken (even if it's not this PR's fault). We'll probably need to come up with some kind of contingency, since I don't think it's possible to get more memory from Travis (even if you pay them).

garyb commented:
The code we have under `test` is basically self contained, isn't it - by which I mean it doesn't depend on anything in `src`?

If so, it seems like we should be able to build the tests separately - we do bundle them separately currently, but `psc` builds everything together as far as I can see.

jonsterling commented:
@garyb So memory usage should improve if we built the tests separately?

garyb commented:
I'm not sure, that's just going from Maxim's suggestion: https://github.com/slamdata/slamdata/pull/490#issuecomment-148806015

That should buy us some time at least, but I guess it will become a problem again at some point if the compiler behaves as badly in the future as it does currently. Phil has been working on that already for 0.7.5, but I think that's mainly the compile performance he's managed to improve.

beckyconning commented:
What about precompiled tests?

cryogenian commented:
+ removed one `fromJust`, added comment to other `fromJust`
+ changed versions from `~` to `^`
+ splitted `make` to `make` and `test-make` 

Now travis is happy :fireworks:

beckyconning commented:
🎉

jonsterling commented:
Nicely done @cryogenian!!
